### ISSUE TITLE
Model training breaks when training on ERA5+IMERG

### ISSUE THREAD

**Issue by mlangguth89** [ecmwf/WeatherGenerator#325]
### What happened?

When training an MTM model on the 6-hourly ERA5 and IMERG anemoi datasets, the training breaks at fairly late stages (at around epoch 16 with 521 samples per epoch). While the losses decrease beforehand, the `inf`-values are detected in the gradient calculation. Example output:
```
 0: 016 : 00420/00512 : 008612 : loss = 2.6992E-01 (lr=4.94E-05, s/sec=0.325)
 0:     ERA5 : 1.1204E-01       IMERG : 4.2780E-01
 0:
 0: 016 : 00430/00512 : 008622 : loss = 2.7591E-01 (lr=4.93E-05, s/sec=0.329)
 0:     ERA5 : 1.2190E-01       IMERG : 4.2992E-01
 0:
 0: > /p/project1/hclimrep/langguth1/WeatherGenerator_ownfork/.venv/lib/python3.12/site-packages/torch/amp/grad_scaler.py(454)step()
 0: -> len(optimizer_state["found_inf_per_device"]) > 0
```

### What are the steps to reproduce the bug?

Run MTM pretraining on ERA5 and IMERG data (both anemoi datasets providing 6h-data).
- IMERG: `imerg-nasa-grib-n320-1998-2024-6h-v1.zarr`
- ERA5: `aifs-ea-an-oper-0001-mars-o96-1979-2022-6h-v6.zarr`
The streams are configured as follows:
ERA5:
```
ERA5 :
  type : anemoi
  filenames : ['aifs-ea-an-oper-0001-mars-o96-1979-2022-6h-v6.zarr']
  loss_weight : 1.
  diagnostic : False
  masking_rate : 0.6
  masking_rate_none : 0.05
  token_size : 32
  embed :
    net : transformer
    num_tokens : 1
    num_heads : 8
    dim_embed : 512
    num_blocks : 2
  embed_target_coords :
    net : linear
```
IMERG:
```
IMERG :
  type : anemoi
  filenames : ['imerg-nasa-grib-n320-1998-2024-6h-v1.zarr']
  loss_weight : 1.
  diagnostic : False
  masking_rate : 0.6
  masking_rate_none : 0.05
  token_size : 256
  embed :
    net : transformer
    num_tokens : 1
    num_heads : 8
    dim_embed : 128
    num_blocks : 2
  embed_target_coords :
    net : linear
    dim_embed : 128
  target_readout :
    type : 'obs_value'  # token or obs_value
    num_layers : 2
    num_heads : 4
  pred_head :
    ens_size : 1
    num_layers : 1
```
The training config is as follows:
```

streams_directory: "/p/home/jusers/langguth1/juwels/WeatherGenerator_ownfork/config/streams/streams_era5+imerg"

embed_orientation: "channels"
embed_local_coords: True
embed_centroids_local_coords: False
embed_size_centroids: 0
embed_unembed_mode: "block"

target_cell_local_prediction: True

ae_local_dim_embed: 2048
ae_local_num_blocks: 2
ae_local_num_heads: 16
ae_local_dropout_rate: 0.1
ae_local_with_qk_lnorm: True

ae_local_num_queries: 2
ae_local_queries_per_cell: False
ae_adapter_num_heads: 16
ae_adapter_embed: 128
ae_adapter_with_qk_lnorm: True
ae_adapter_with_residual: True
ae_adapter_dropout_rate: 0.1

ae_global_dim_embed: 2048
ae_global_num_blocks: 10
ae_global_num_heads: 32
ae_global_dropout_rate: 0.1
ae_global_with_qk_lnorm: True
ae_global_att_dense_rate: 0.2
ae_global_block_factor: 64
ae_global_mlp_hidden_factor: 2

pred_adapter_kv: False
pred_self_attention: True
pred_dyadic_dims: False
pred_mlp_adaln: True

# number of steps offset applied to first target window; if set to zero and forecast_steps=0 then
# one is training an auto-encoder
forecast_offset : 0
forecast_delta_hrs: 0
forecast_steps: 0
forecast_policy: null
forecast_freeze_model: False
forecast_att_dense_rate: 0.25
fe_num_blocks: 0
fe_num_heads: 16
fe_dropout_rate: 0.1
fe_with_qk_lnorm: True

healpix_level: 5

with_mixed_precision: True
with_flash_attention: True
compile_model: False
with_fsdp: True

loss_fcts:
  -
    - "mse"
    - 1.0
loss_fcts_val:
  -
    - "mse"
    - 1.0

batch_size: 1
batch_size_validation: 1

# training mode: "forecast" or "masking" (masked token modeling)
# for "masking" to train with auto-encoder mode, forecast_offset should be 0
training_mode: "masking"
# masking rate when training mode is "masking"; ignored in foreacast mode
masking_rate: 0.6
# sample the masking rate (with normal distribution centered at masking_rate)
masking_rate_sampling: True
# sample a subset of all target points, useful e.g. to reduce memory requirements
sampling_rate_target: 1.0

num_epochs: 36
samples_per_epoch: 512
samples_per_validation: 512
shuffle: True

lr_scaling_policy: "sqrt"
lr_start: 0.000001
lr_max: 0.00003
lr_final_decay: 0.000001
lr_final: 0.0
r_steps_warmup: 256
lr_steps_cooldown: 4096
lr_policy_warmup: "cosine"
lr_policy_decay: "linear"
lr_policy_cooldown: "linear"

grad_clip: 5.0
weight_decay: 0.1
norm_type: "LayerNorm"
nn_module: "te"


start_date: 199801010000
end_date: 202012310000
start_date_val: 202101010000
end_date_val: 202112310000
len_hrs: 6
step_hrs: 6
input_window_steps: 1

val_initial: False

loader_num_workers: 8
log_validation: 0

istep: 0
run_history: []

desc: ""
data_loader_rng_seed: ???
run_id: ???

# Parameters for logging/printing in the training loop
train_log:
  # The period to log metrics (in number of batch steps)
  log_interval: 20
```

### Version

current develop (2025-06-12)

### Platform (OS and architecture)

Juwels Booster

### Relevant log output

```shell
0: 016 : 00420/00512 : 008612 : loss = 2.6992E-01 (lr=4.94E-05, s/sec=0.325)
 0:     ERA5 : 1.1204E-01       IMERG : 4.2780E-01
 0:
 0: 016 : 00430/00512 : 008622 : loss = 2.7591E-01 (lr=4.93E-05, s/sec=0.329)
 0:     ERA5 : 1.2190E-01       IMERG : 4.2992E-01
 0:
 0: > /p/project1/hclimrep/langguth1/WeatherGenerator_ownfork/.venv/lib/python3.12/site-packages/torch/amp/grad_scaler.py(454)step()
 0: -> len(optimizer_state["found_inf_per_device"]) > 0
```

### Organisation

JSC

**Comment by mlangguth89** [comment]
It will be tested if this issue is resolved with the changes to the target coordinate embedding introduced in #356.
